### DISCUSSION TITLE
Uninstalled package keeps coming back (bloop)

### DISCUSSION THREAD

**Discussion by jrduncans** [Homebrew/discussions#77] [Everyday usage] [answered]
I've run various forms of `brew uninstall bloop` and it often seems gone, but it seems to keep coming back.

I'll notice it come back because it will try to upgrade and fail with permission issues (either deleting the old version or creating the directory for the new version -- the latter may have only started happening after I `sudo rm -rf /usr/local/Cellar/bloop` at some point to try to clear out all references.

```
==> Upgrading 2 outdated packages:
scalacenter/bloop/bloop 1.4.4
git 2.28.0 -> 2.29.0
==> Upgrading scalacenter/bloop/bloop -> 1.4.4
==> Downloading https://github.com/scalacenter/bloop/releases/download/v1.4.4/bash-completions
Already downloaded: /Users/sdunc4/Library/Caches/Homebrew/downloads/bda989952276aaa86d22e7a95cb24d0e205a603370c4da28a50d71f0f870175a--bash-completions
==> Downloading https://github.com/scalacenter/bloop/releases/download/v1.4.4/zsh-completions
Already downloaded: /Users/sdunc4/Library/Caches/Homebrew/downloads/f7bde3ed0b4cb3edfb383bf2b4e5ffa7d265d4d227fd70e40e03cf54525b68f7--zsh-completions
==> Downloading https://github.com/scalacenter/bloop/releases/download/v1.4.4/fish-completions
Already downloaded: /Users/sdunc4/Library/Caches/Homebrew/downloads/60eee26b73dd69ceeb1649723980d613d9eff13594e2fe1677b027b7ccd52948--fish-completions
==> Downloading https://github.com/scalacenter/bloop/releases/download/v1.4.4/bloop-coursier.json
Already downloaded: /Users/sdunc4/Library/Caches/Homebrew/downloads/a0e1bf997af5a38cfa1dc04b434b6453829fb2e4628fe63b6df196a49ce94ce3--bloop-coursier.json
Error: An exception occurred within a child process:
  Errno::EACCES: Permission denied @ dir_s_mkdir - /usr/local/Cellar/bloop/1.4.4
```

This seems weird and specific, so haven't been sure it's worth filing a bug, but if anyone can advise on how to make sure I remove it so it never comes back, I'm all ears.

**Comment by jrduncans** [comment, marked as answer]
I tried `brew uninstall scalacenter/bloop/bloop`, it said it couldn't remove `/usr/local/Cellar/bloop/1.3.4`, so I ran `sudo rm -rf /usr/local/Cellar/bloop/1.3.4`. Now `brew uninstall scalacenter/bloop/bloop` says 'Error: No installed keg or cask with the name "scalacenter/bloop/bloop"'. Same for `brew uninstall bloop`. However `brew list` still shows `bloop` in the list.

`brew uninstall bloop -f` said "Uninstalling bloop... (64B)", but `brew list` still shows `bloop` in the list. Same results for `brew uninstall -f scalacenter/bloop/bloop`. 

`sudo rm -rf /usr/local/Cellar/bloop/` finally makes `brew list` not show `bloop`, but I've been here before and it will come back in the next couple of weeks somehow.

**Reply by jrduncans** [reply]
Perhaps running `brew untap scalacenter/bloop` will prevent it from coming back? We shall see...

**Reply by SMillerDev** [reply]
You can also try `brew cleanup` and see if that helps.

**Reply by jrduncans** [reply]
I always run brew cleanup afterwards, so that unfortunately hasn't helped. Thanks though.

**Reply by jrduncans** [reply]
Seems like `brew untap scalacenter/bloop` did it, so many it was something wrong with that tap.

**Comment by dbersan** [comment]
It's mind blowing that removing a package is so hard.

You do `brew uninstall XXXX`

and then `which XXX` still shows a path. Why?

**Reply by SMillerDev** [reply]
Is it because your shell caches the answer?

**Reply by gromgit** [reply]
If your shell is zsh, `which XXX` is a builtin command that consults its internal hash table for `XXX`. You need to run `hash -r` to clear it, then `which XXX` should say `XXX not found`.
